### PR TITLE
Storing the annotation state in `with_annotation`

### PR DESCRIPTION
**Pull request by Ig-dolci**
## Fixes Issue:
- https://github.com/dolfin-adjoint/pyadjoint/issues/111

## Depends on:

- https://github.com/dolfin-adjoint/dolfin-adjoint/pull/12

### CODE REVIEW

File: pyadjoint/tape.py [hunk 1]
```diff
@@ -125,7 +130,7 @@ def annotate_tape(kwargs=None):
 
     # TODO: Consider if there is any scenario where one would want the keyword to have
     # precedence over the global flag.
-    if _stop_annotating > 0:
+    if _annotation_enabled is False:
```
Comment (connorjward) [inline pyadjoint/tape.py:133]: Use `if not _annotation_enabled`.

**Review by connorjward — CHANGES_REQUESTED** [review]

File: pyadjoint/tape.py [hunk 2]
```diff
@@ -82,12 +82,17 @@ class stop_annotating(object):
 
     def __init__(self, modifies=None):
         self.modifies = modifies
+        self._orig_annotation_enabled = False
```
[the remark this answers is not in the document]
Comment (Ig-dolci) [reply pyadjoint/tape.py]: done.

File: pyadjoint/tape.py (hunk 1 above)
Comment (Ig-dolci) [reply pyadjoint/tape.py]: done.

File: pyadjoint/tape.py [hunk 3]
```diff
@@ -125,7 +130,7 @@ def annotate_tape(kwargs=None):
 
     # TODO: Consider if there is any scenario where one would want the keyword to have
     # precedence over the global flag.
-    if _stop_annotating > 0:
+    if _annotation_enabled is False:
         return False
 
     return annotate
```
[the remark this answers is not in the document]
Comment (Ig-dolci) [reply pyadjoint/tape.py:136]: This modification breaks the firedrake tests. I think `annotate` needs to be defined anyway.

[the remark this answers is not in the document]
Comment (dham) [reply pyadjoint/tape.py:135]: I agree. @connorjward's change would change the behaviour of the function. For better or worse, the function currently has the side effect of always removing the `annotate` keyword.

File: pyadjoint/tape.py [hunk 4]
```diff
@@ -9,22 +9,22 @@
...
+    _annotation_enabled = False
 
 
 def continue_annotation():
-    global _stop_annotating
-    _stop_annotating -= 1
-    return _stop_annotating <= 0
+    global _annotation_enabled
+    _annotation_enabled = True
+    return _annotation_enabled
```
[the remark this answers is not in the document]
Comment (dham) [reply pyadjoint/tape.py:27]: Let's not tempt fate.

File: .circleci/config.yml [hunk 5]
```diff
@@ -59,7 +59,7 @@ jobs:
       - run:
           name: Install dolfin-adjoint
           command: |
-            git clone https://github.com/dolfin-adjoint/dolfin-adjoint ~/dolfin-adjoint
+            git clone --branch dham/default_annotate_on https://github.com/dolfin-adjoint/dolfin-adjoint ~/dolfin-adjoint
```
Comment (dham) [inline .circleci/config.yml:62]:
```suggestion
            git clone https://github.com/dolfin-adjoint/dolfin-adjoint ~/dolfin-adjoint
```

**Pull request merged**